File: 6. Binary Search/875. Koko Eating Bananas/Solution.py
from typing import List
import math

class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        max_speed = max(piles)
        min_speed = 1

        while (min_speed < max_speed):
            middle = min_speed + (max_speed - min_speed) // 2
            eating_time = self.eat_bananas(piles, middle)

            if eating_time <= h:
                max_speed = middle
            else:
                min_speed = middle + 1
        
        return min_speed
    
    def eat_bananas(self, piles, speed):
        time = 0

        for pile in piles:
            time += math.ceil(pile / speed)
        
        return time

File: 6. Binary Search/875. Koko Eating Bananas/Solution.cpp
#include <vector>

class Solution {
public:
    int minEatingSpeed(std::vector<int>& piles, int h) {
        int min = 1;
        int max = *std::max_element(piles.begin(), piles.end());

        while (min < max)
        {
            int middle = min + (max - min) / 2;
            int eatingTime = eatBananas(piles, middle);
            if (eatingTime <= h) max = middle;
            else min = middle + 1;
        }

        return min;
    }

    int eatBananas(std::vector<int>& piles, int speed)
    {
        int time = 0;
        for (int pile : piles) time += (pile + speed - 1) / speed;
        return time;
    }
};